File: backend/app/modules/assistant/app_context.py
"""Bounded application state supplied by a Nova surface for one Nove turn."""

from __future__ import annotations

import json
import re
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator, model_validator

from app.common.sql_guard import CredentialsRedactionError, redact_sql_credentials

MAX_APP_CONTEXT_BYTES = 24_576
_CAPABILITY_NAME = re.compile(r"^[a-z][a-z0-9_.-]{0,79}$")
_SENSITIVE_TEXT = re.compile(
    r"password|passwd|passphrase|secret|token|credential|"
    r"api[\s._-]*key|access[\s._-]*key|private[\s._-]*key|"
    r"account[\s._-]*key|authorization|authentication|bearer|"
    r"identified\s+by|cookie",
    re.IGNORECASE,
)
_URL_USERINFO = re.compile(r"://[^\s/@]+:[^\s/@]+@")


def _contains_sensitive_text(value: str) -> bool:
    from app.modules.assistant.tools.redaction import is_credential_value

    return bool(
        _SENSITIVE_TEXT.search(value)
        or _URL_USERINFO.search(value)
        or is_credential_value(value)
    )


def _safe_text(value: str) -> str:
    if _contains_sensitive_text(value):
        return "***"
    try:
        redacted = redact_sql_credentials(value)
    except CredentialsRedactionError:
        return "***"
    return "***" if redacted != value else value


def _safe_metadata(value: Any, depth: int = 0, *, sql_text: bool = False) -> Any:
    from app.modules.assistant.tools.redaction import is_credential_column

    if depth > 4:
        return None
    if isinstance(value, dict):
        return {
            str(key)[:80]: _safe_metadata(
                item, depth + 1, sql_text=str(key).casefold() == "sql"
            )
            for key, item in list(value.items())[:32]
            if not is_credential_column(str(key))
            and not _contains_sensitive_text(str(key))
        }
    if isinstance(value, list):
        return [_safe_metadata(item, depth + 1) for item in value[:24]]
    if isinstance(value, str):
        return _safe_text(value)[:4000 if sql_text else 2048]
    if isinstance(value, bool | int | float) or value is None:
        return value
    return None


class _ContextPart(BaseModel):
    model_config = {"extra": "ignore", "populate_by_name": True}

    @model_validator(mode="after")
    def scrub_free_form_strings(self) -> _ContextPart:
        for field_name in type(self).model_fields:
            value = getattr(self, field_name)
            if isinstance(value, str):
                setattr(self, field_name, _safe_text(value))
        return self


class AppSurface(_ContextPart):
    id: str = Field(min_length=1, max_length=128)
    route: str = Field(min_length=1, max_length=512)
    title: str | None = Field(default=None, max_length=256)
    version: int | None = Field(default=None, ge=0)


class AppEntity(_ContextPart):
    type: str = Field(min_length=1, max_length=80)
    id: str | None = Field(default=None, max_length=256)
    name: str | None = Field(default=None, max_length=256)
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("metadata", mode="after")
    @classmethod
    def scrub_metadata(cls, value: dict[str, Any]) -> dict[str, Any]:
        return _safe_metadata(value)


class AppSelection(_ContextPart):
    type: str | None = Field(default=None, max_length=80)
    ids: list[str] = Field(default_factory=list, max_length=32)
    text: str | None = Field(default=None, max_length=4096)
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("ids")
    @classmethod
    def check_ids(cls, values: list[str]) -> list[str]:
        if any(len(value) > 256 for value in values):
            raise ValueError("Selection ids must be at most 256 characters")
        return [_safe_text(value) for value in values]

    @field_validator("text")
    @classmethod
    def scrub_text(cls, value: str | None) -> str | None:
        return _safe_text(value) if value is not None else None

    @field_validator("metadata")
    @classmethod
    def scrub_metadata(cls, value: dict[str, Any]) -> dict[str, Any]:
        return _safe_metadata(value)


class AppCursor(_ContextPart):
    line: int | None = Field(default=None, ge=0)
    column: int | None = Field(default=None, ge=0)


class AppEditor(_ContextPart):
    document_id: str | None = Field(default=None, alias="documentId", max_length=256)
    language: str | None = Field(default=None, max_length=40)
    selected_text: str | None = Field(default=None, alias="selectedText", max_length=4096)
    cursor: AppCursor | None = None
    dirty: bool | None = None

    @field_validator("selected_text")
    @classmethod
    def scrub_selected_text(cls, value: str | None) -> str | None:
        return _safe_text(value) if value is not None else None


class AppResultColumn(_ContextPart):
    name: str = Field(min_length=1, max_length=128)
    type: str | None = Field(default=None, max_length=80)


class AppExecution(_ContextPart):
    type: str | None = Field(default=None, max_length=80)
    execution_id: str | None = Field(default=None, alias="executionId", max_length=128)
    status: Literal["idle", "running", "success", "error"] | None = None
    error_code: str | None = Field(default=None, alias="errorCode", max_length=128)
    error_message: str | None = Field(default=None, alias="errorMessage", max_length=2048)
    elapsed_ms: float | None = Field(default=None, alias="elapsedMs", ge=0)
    row_count: int | None = Field(default=None, alias="rowCount", ge=0)
    result_schema: list[AppResultColumn] = Field(
        default_factory=list, alias="resultSchema", max_length=64
    )

    @field_validator("error_message")
    @classmethod
    def scrub_error_message(cls, value: str | None) -> str | None:
        return _safe_text(value) if value is not None else None


class AppView(_ContextPart):
    active_tab: str | None = Field(default=None, alias="activeTab", max_length=128)
    filters: dict[str, Any] = Field(default_factory=dict)
    search: str | None = Field(default=None, max_length=256)
    sort: Any = None

    @field_validator("filters", "sort")
    @classmethod
    def scrub_values(cls, value: Any) -> Any:
        return _safe_metadata(value)


class AppDomain(_ContextPart):
    database: str | None = Field(default=None, max_length=128)
    schema_name: str | None = Field(default=None, alias="schema", max_length=128)
    role: str | None = Field(default=None, max_length=128)
    semantic_model: str | None = Field(default=None, alias="semanticModel", max_length=128)
    provider_id: str | None = Field(default=None, alias="providerId", max_length=128)


class NoveApplicationEvent(_ContextPart):
    id: str = Field(min_length=1, max_length=128)
    timestamp: datetime
    source: Literal["assistant", "surface", "execution", "user"]
    type: str = Field(min_length=1, max_length=80)
    surface_id: str | None = Field(default=None, alias="surfaceId", max_length=128)
    correlation_id: str | None = Field(default=None, alias="correlationId", max_length=128)
    artifact_id: str | None = Field(default=None, alias="artifactId", max_length=128)
    execution_id: str | None = Field(default=None, alias="executionId", max_length=128)
    status: Literal["success", "failure"] | None = None
    payload: dict[str, Any] = Field(default_factory=dict)

    @field_validator("payload")
    @classmethod
    def scrub_payload(cls, value: dict[str, Any]) -> dict[str, Any]:
        return _safe_metadata(value)


class NoveAppContext(_ContextPart):
    version: Literal[1]
    surface: AppSurface
    entity: AppEntity | None = None
    selection: AppSelection | None = None
    editor: AppEditor | None = None
    execution: AppExecution | None = None
    view: AppView | None = None
    domain: AppDomain | None = None
    capabilities: list[str] = Field(default_factory=list, max_length=32)
    events: list[NoveApplicationEvent] = Field(default_factory=list, max_length=12)

    @model_validator(mode="before")
    @classmethod
    def bound_raw_envelope(cls, value: Any) -> Any:
        try:
            size = len(json.dumps(value, ensure_ascii=False, default=str).encode("utf-8"))
        except (TypeError, ValueError) as exc:
            raise ValueError("Application context must be serializable") from exc
        if size > MAX_APP_CONTEXT_BYTES:
            raise ValueError("Application context exceeds 24 KB")
        return value

    @field_validator("capabilities")
    @classmethod
    def check_capabilities(cls, values: list[str]) -> list[str]:
        values = [value for value in values if not _contains_sensitive_text(value)]
        if any(not _CAPABILITY_NAME.fullmatch(value) for value in values):
            raise ValueError("Application context has an invalid capability name")
        return list(dict.fromkeys(values))

    def current_events(self) -> list[NoveApplicationEvent]:
        current_document = self.editor.document_id if self.editor else None
        current_entity = self.entity.id if self.entity else None
        visible: list[NoveApplicationEvent] = []
        for event in self.events:
            if event.surface_id is not None and event.surface_id != self.surface.id:
                continue
            document_id = event.payload.get("documentId")
            if document_id and document_id != current_document:
                continue
            entity_id = event.payload.get("entityId")
            if entity_id and entity_id != current_entity:
                continue
            visible.append(event)
        return visible[-8:]

    def prompt_data(self) -> dict[str, Any]:
        data = self.model_dump(by_alias=True, mode="json", exclude_none=True)
        data["events"] = [
            event.model_dump(by_alias=True, mode="json", exclude_none=True)
            for event in self.current_events()
        ]
        return data


_DEICTIC = re.compile(
    r"\b(this|that|it|here|these|current|selected|last|ini|itu|terakhir)\b", re.I
)


def resolve_app_references(request: str, app: NoveAppContext | None) -> dict[str, Any]:
    """Give the model bounded references from the current surface, not old turns."""
    if app is None or not _DEICTIC.search(request):
        return {}
    result: dict[str, Any] = {"surfaceId": app.surface.id}
    selection = app.selection
    if selection and (selection.ids or selection.text):
        result["selection"] = selection.model_dump(by_alias=True, exclude_none=True)
    if app.entity is not None:
        result["entity"] = app.entity.model_dump(by_alias=True, exclude_none=True)
    if app.execution is not None and app.execution.status in {"error", "success"}:
        result["execution"] = app.execution.model_dump(by_alias=True, exclude_none=True)
    elif failure := next(
        (event for event in reversed(app.current_events()) if event.type == "query_failed"),
        None,
    ):
        result["lastFailure"] = failure.model_dump(by_alias=True, exclude_none=True)
    if app.editor is not None and app.editor.selected_text:
        result["editorSelection"] = app.editor.selected_text
    return result

File: frontend/src/features/assistant/app-context.ts
/** The bounded application state Nove receives with each turn. */
export type NoveAppContext = {
  version: 1;
  surface: { id: string; route: string; title?: string | null };
  entity?: {
    type: string;
    id?: string | null;
    name?: string | null;
    metadata?: Record<string, unknown>;
  };
  selection?: {
    type?: string | null;
    ids?: string[];
    text?: string | null;
    metadata?: Record<string, unknown>;
  };
  editor?: {
    documentId?: string | null;
    language?: string | null;
    selectedText?: string | null;
    cursor?: { line?: number; column?: number };
    dirty?: boolean;
  };
  execution?: {
    type?: string | null;
    executionId?: string | null;
    status?: "idle" | "running" | "success" | "error";
    errorCode?: string | null;
    errorMessage?: string | null;
    elapsedMs?: number | null;
    rowCount?: number | null;
    resultSchema?: Array<{ name: string; type?: string | null }>;
  };
  view?: {
    activeTab?: string | null;
    filters?: Record<string, unknown>;
    search?: string | null;
    sort?: string | null;
  };
  domain?: {
    database?: string | null;
    schema?: string | null;
    role?: string | null;
    semanticModel?: string | null;
    providerId?: string | null;
  };
  capabilities: string[];
  events?: NoveApplicationEvent[];
};

export type NoveSurfaceContext = Partial<
  Omit<NoveAppContext, "version" | "surface" | "capabilities" | "events">
>;

export type NoveApplicationEvent = {
  id: string;
  timestamp: string;
  source: "assistant" | "surface" | "execution" | "user";
  type: string;
  surfaceId?: string;
  correlationId?: string;
  artifactId?: string;
  executionId?: string;
  status?: "success" | "failure";
  payload?: Record<string, unknown>;
};

export type NoveEventInput = Omit<NoveApplicationEvent, "id" | "timestamp">;

const MAX_CONTEXT_BYTES = 16_384;
const MAX_EVENTS = 8;
const SENSITIVE_TEXT =
  /password|passwd|passphrase|secret|token|credential|api[\s._-]*key|access[\s._-]*key|private[\s._-]*key|account[\s._-]*key|authorization|authentication|bearer|identified\s+by|cookie/i;
const CREDENTIAL_VALUE =
  /\b(?:AKIA|ASIA|AGPA|AIDA|AROA|AIPA|ANPA|ANVA)[A-Z0-9]{16}\b|-----BEGIN (?:RSA |EC |DSA |OPENSSH |PGP )?PRIVATE KEY-----|\b(?:sk|gsk|xai|ghp|gho|ghu|ghs|ghr)[-_][A-Za-z0-9_-]{16,}\b|\bxox[baprs][-_][A-Za-z0-9-]{10,}\b|\beyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\b|:\/\/[^\s/@]+:[^\s/@]+@/i;

function containsSensitiveText(value: string): boolean {
  return SENSITIVE_TEXT.test(value) || CREDENTIAL_VALUE.test(value);
}

function clip(value: unknown, limit = 240): string | undefined {
  if (typeof value !== "string") return undefined;
  // Inspect before clipping: otherwise a credential after the limit survives in a shorter field.
  return containsSensitiveText(value) ? "[REDACTED]" : value.slice(0, limit);
}

function clipSensitiveText(value: unknown, limit: number): string | undefined {
  return clip(value, limit);
}

function scalarRecord(
  value: unknown,
  maxEntries = 12,
  allowSql = false,
): Record<string, unknown> | undefined {
  if (!value || typeof value !== "object" || Array.isArray(value))
    return undefined;
  const result: Record<string, unknown> = {};
  for (const [key, entry] of Object.entries(value).slice(0, maxEntries)) {
    if (containsSensitiveText(key)) continue;
    const safeKey = clip(key, 64);
    if (!safeKey) continue;
    if (typeof entry === "string") {
      if (allowSql && key === "sql") {
        if (!containsSensitiveText(entry)) result.sql = clip(entry, 4_000);
      } else result[safeKey] = clip(entry);
    } else if (typeof entry === "number" && Number.isFinite(entry))
      result[safeKey] = entry;
    else if (typeof entry === "boolean" || entry === null)
      result[safeKey] = entry;
  }
  return Object.keys(result).length ? result : undefined;
}

/** Only named fields cross the model boundary; unknown surface state is discarded. */
export function boundAppContext(context: NoveAppContext): NoveAppContext {
  const current: NoveAppContext = {
    version: 1,
    surface: {
      id: clip(context.surface.id, 96) || "nova",
      route: clip(context.surface.route, 256) || "/",
      title: clip(context.surface.title, 120),
    },
    capabilities: [...new Set(context.capabilities)]
      .filter((value) => !containsSensitiveText(value))
      .slice(0, 32)
      .map((value) => clip(value, 96) || "")
      .filter(Boolean),
  };
  if (context.entity)
    current.entity = {
      type: clip(context.entity.type, 64) || "unknown",
      id: clip(context.entity.id, 160),
      name: clip(context.entity.name, 160),
      metadata: scalarRecord(context.entity.metadata),
    };
  if (context.selection)
    current.selection = {
      type: clip(context.selection.type, 64),
      ids: context.selection.ids?.slice(0, 20).map((id) => clip(id, 160) || ""),
      text: clipSensitiveText(context.selection.text, 2_000),
      metadata: scalarRecord(context.selection.metadata),
    };
  if (context.editor)
    current.editor = {
      documentId: clip(context.editor.documentId, 160),
      language: clip(context.editor.language, 32),
      selectedText: clipSensitiveText(context.editor.selectedText, 2_000),
      cursor: context.editor.cursor
        ? {
            line: Number.isSafeInteger(context.editor.cursor.line)
              ? context.editor.cursor.line
              : undefined,
            column: Number.isSafeInteger(context.editor.cursor.column)
              ? context.editor.cursor.column
              : undefined,
          }
        : undefined,
      dirty: context.editor.dirty === true,
    };
  if (context.execution)
    current.execution = {
      type: clip(context.execution.type, 64),
      executionId: clip(context.execution.executionId, 160),
      status: context.execution.status,
      errorCode: clip(context.execution.errorCode, 80),
      errorMessage: clipSensitiveText(context.execution.errorMessage, 1_500),
      elapsedMs:
        typeof context.execution.elapsedMs === "number" &&
        Number.isFinite(context.execution.elapsedMs)
          ? context.execution.elapsedMs
          : undefined,
      rowCount:
        typeof context.execution.rowCount === "number" &&
        Number.isFinite(context.execution.rowCount)
          ? context.execution.rowCount
          : undefined,
      resultSchema: context.execution.resultSchema
        ?.slice(0, 30)
        .map((item) => ({
          name: clip(item.name, 96) || "",
          type: clip(item.type, 64),
        })),
    };
  if (context.view)
    current.view = {
      activeTab: clip(context.view.activeTab, 80),
      filters: scalarRecord(context.view.filters),
      search: clip(context.view.search, 240),
      sort: clip(context.view.sort, 80),
    };
  if (context.domain)
    current.domain = {
      database: clip(context.domain.database, 120),
      schema: clip(context.domain.schema, 120),
      role: clip(context.domain.role, 120),
      semanticModel: clip(context.domain.semanticModel, 160),
      providerId: clip(context.domain.providerId, 160),
    };
  current.events = context.events?.slice(-MAX_EVENTS).map((event) => ({
    id: clip(event.id, 96) || "",
    timestamp: clip(event.timestamp, 40) || "",
    source: event.source,
    type: clip(event.type, 80) || "unknown",
    surfaceId: clip(event.surfaceId, 96),
    correlationId: clip(event.correlationId, 160),
    artifactId: clip(event.artifactId, 160),
    executionId: clip(event.executionId, 160),
    status: event.status,
    payload: scalarRecord(event.payload, 16, true),
  }));
  // Trim the least important data first. Surface and capability identity remain.
  while (
    new TextEncoder().encode(JSON.stringify(current)).length > MAX_CONTEXT_BYTES
  ) {
    if (current.events?.length) current.events.shift();
    else if (current.selection?.text)
      current.selection.text = current.selection.text.slice(
        0,
        Math.floor(current.selection.text.length / 2),
      );
    else if (current.editor?.selectedText)
      current.editor.selectedText = current.editor.selectedText.slice(
        0,
        Math.floor(current.editor.selectedText.length / 2),
      );
    else if (current.execution?.errorMessage)
      current.execution.errorMessage = current.execution.errorMessage.slice(
        0,
        Math.floor(current.execution.errorMessage.length / 2),
      );
    else if (current.entity?.metadata) delete current.entity.metadata;
    else if (current.selection?.metadata) delete current.selection.metadata;
    else if (current.view?.filters) delete current.view.filters;
    else if (current.execution?.resultSchema)
      delete current.execution.resultSchema;
    else if (current.selection) delete current.selection;
    else if (current.editor) delete current.editor;
    else if (current.execution) delete current.execution;
    else break;
  }
  return current;
}

let eventSequence = 0;
export function createApplicationEvent(
  input: NoveEventInput,
): NoveApplicationEvent {
  eventSequence += 1;
  return {
    ...input,
    id: `${Date.now()}-${eventSequence}`,
    timestamp: new Date().toISOString(),
  };
}
